File: tythan-code/src/core/rules.ts
/**
 * Project rules files — the .cursorrules idea: a plain-text file at the
 * workspace root whose contents are appended to the system prompt, so a
 * project can pin its conventions ("always use pnpm", "tests live in
 * tests/", ...) without the user repeating them every session.
 *
 * Read fresh on every turn (it's one tiny file read), so edits take effect
 * immediately with no watcher wiring.
 */

import * as fs from "node:fs";
import * as path from "node:path";

// First match wins. `.cursorrules` is supported so an existing Cursor
// project's rules keep working unchanged.
export const RULES_FILENAMES = [".tythanrules", ".cursorrules", "AGENTS.md"];

export const MAX_RULES_CHARS = 8_000;

export interface RulesText {
  source: string;
  text: string;
}

export function loadRulesText(workspaceRoot: string): RulesText | undefined {
  for (const name of RULES_FILENAMES) {
    const file = path.join(workspaceRoot, name);
    let text: string;
    try {
      if (!fs.statSync(file).isFile()) {
        continue;
      }
      text = fs.readFileSync(file, "utf-8");
    } catch {
      continue;
    }
    const trimmed = text.trim();
    if (!trimmed) {
      continue;
    }
    return {
      source: name,
      text: trimmed.length > MAX_RULES_CHARS ? trimmed.slice(0, MAX_RULES_CHARS) + "\n... [rules truncated]" : trimmed,
    };
  }
  return undefined;
}

File: tythan/rules.py
# TythanAI Security Platform — Community Edition
# Copyright (c) 2026 TythanAI Labs
# Licensed under the Business Source License 1.1 (see LICENSE).

"""
tythan/rules.py — project rules files.

Instructions in `.tythanrules`, `.cursorrules` (existing Cursor projects
keep working unchanged) or `AGENTS.md` at the workspace root are appended
to the system prompt on every turn. Re-read each turn, so edits apply on
the next message.
"""
from __future__ import annotations

from pathlib import Path

RULES_FILENAMES = (".tythanrules", ".cursorrules", "AGENTS.md")
MAX_RULES_CHARS = 16_000


def load_rules(workspace_root: Path) -> str:
    """Return the first non-empty rules file's content (precedence order),
    truncated to MAX_RULES_CHARS."""
    for name in RULES_FILENAMES:
        fp = workspace_root / name
        try:
            if not fp.is_file():
                continue
            text = fp.read_text("utf-8", errors="replace").strip()
        except OSError:
            continue
        if not text:
            continue
        if len(text) > MAX_RULES_CHARS:
            text = text[:MAX_RULES_CHARS] + "\n…(rules truncated)"
        return f"\n\n# Project rules (from {name})\n{text}"
    return ""
